src/metabase/models/app.clj

```diff
@@ -41,3 +42,56 @@
           (cond-> coll
             app-id (assoc :app_id app-id)))))
     collections))
+
+(defn- app-cards [app]
+  (->> (db/query {:union
+                  [{:select [:c.*]
+                    :from [[:report_card :c]]
+                    :where [:and
+                            [:= :c.collection_id (:collection_id app)]]}
+                   {:select [:c.*]
+                    :from [[:report_card :c]]
+                    :join [[:report_dashboardcard :dc] [:= :dc.card_id :c.id]
+                           [:report_dashboard :d] [:= :d.id :dc.dashboard_id]]
+                    :where [:and
+                            [:= :d.collection_id (:collection_id app)]]}]})
+       (db/do-post-select 'Card)))
+
+(defn- parse-source-query-id
+  "Return the ID of the card used as source table, if applicable; otherwise return `nil`."
+  [source-table]
+  (when (string? source-table)
+    (when-let [[_ card-id-str] (re-matches #"^card__(\d+$)" source-table)]
+      (parse-long card-id-str))))
+
+(defn- collect-model-ids
+  "Return a sequence of model ids referenced in the MBQL query `mbql-form`."
+  [mbql-form]
+  (let [ids (java.util.HashSet.)
+        walker (fn [form]
+                 (when (map? form)
+                   ;; model references in native queries
+                   (when-let [card-id (:card-id form)]
+                     (when (int? card-id)
+                       (.add ids card-id)))
+                   ;; source tables (possibly in joins)
+                   (when-let [card-id (parse-source-query-id (:source-table form))]
```

Comment:
There is a `models.query/query->database-and-table-ids` this could live in a similar place.